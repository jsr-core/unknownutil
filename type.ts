/**
 * A type predicate function.
 *
 * ```ts
 * import { as, is, type Predicate } from "@core/unknownutil";
 *
 * type Person = {
 *   name: string;
 *   age: number;
 *   address?: string;
 * };
 * const isPerson = is.ObjectOf({
 *   name: is.String,
 *   age: is.Number,
 *   address: as.Optional(is.String),
 * }) satisfies Predicate<Person>;
 * ```
 */
export type Predicate<T> = (x: unknown) => x is T;

/**
 * A type predicated by {@linkcode Predicate<T>}.
 *
 * ```ts
 * import { as, is, type PredicateType } from "@core/unknownutil";
 *
 * const isPerson = is.ObjectOf({
 *   name: is.String,
 *   age: is.Number,
 *   address: as.Optional(is.String),
 * });
 *
 * type Person = PredicateType<typeof isPerson>;
 * // type Person = {
 * //   name: string;
 * //   age: number;
 * //   address?: string;
 * // };
 * ```
 */
export type PredicateType<P> = P extends Predicate<infer T> ? T : never;

/**
 * JavaScript primitive types.
 */
export type Primitive =
  | string
  | number
  | bigint
  | boolean
  | null
  | undefined
  | symbol;
