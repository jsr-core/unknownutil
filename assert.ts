import type { Predicate } from "./type.ts";

/**
 * A factory function that generates assertion error messages.
 */
export type AssertMessageFactory = (
  x: unknown,
  pred: Predicate<unknown>,
  name?: string,
) => string;

/**
 * The default factory function used to generate assertion error messages.
 */
export const defaultAssertMessageFactory: AssertMessageFactory = (
  x,
  pred,
  name,
) => {
  const p = pred.name || "anonymous predicate";
  const t = typeof x;
  const v = JSON.stringify(x, null, 2);
  return `Expected ${
    name ?? "a value"
  } that satisfies the predicate ${p}, got ${t}: ${v}`;
};

let assertMessageFactory = defaultAssertMessageFactory;

/**
 * Represents an error that occurs when an assertion fails.
 */
export class AssertError extends Error {
  /**
   * Constructs a new instance.
   * @param message The error message.
   */
  constructor(message?: string) {
    super(message);

    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, AssertError);
    }

    this.name = this.constructor.name;
  }
}

/**
 * Sets the factory function used to generate assertion error messages.
 *
 * ```ts
 * import { is, setAssertMessageFactory } from "@core/unknownutil";
 *
 * setAssertMessageFactory((x, pred) => {
 *   if (pred === is.String) {
 *     return `Expected a string, got ${typeof x}`;
 *   } else if (pred === is.Number) {
 *     return `Expected a number, got ${typeof x}`;
 *   } else if (pred === is.Boolean) {
 *     return `Expected a boolean, got ${typeof x}`;
 *   } else {
 *     return `Expected a value that satisfies the predicate, got ${typeof x}`;
 *   }
 * });
 * ```
 *
 * @param factory The factory function.
 */
export function setAssertMessageFactory(factory: AssertMessageFactory): void {
  assertMessageFactory = factory;
}

/**
 * Asserts that the given value satisfies the provided predicate.
 *
 * It throws {@linkcode AssertError} if the value does not satisfy the predicate.
 *
 * ```ts
 * import { assert, is } from "@core/unknownutil";
 *
 * const a: unknown = "hello";
 * assert(a, is.String);
 * // a is now narrowed to string
 * ```
 *
 * @param x The value to be asserted.
 * @param pred The predicate function to test the value against.
 * @param options Optional configuration for the assertion.
 * @returns The function has a return type of `asserts x is T` to help TypeScript narrow down the type of `x` after the assertion.
 */
export function assert<T>(
  x: unknown,
  pred: Predicate<T>,
  options: { message?: string; name?: string } = {},
): asserts x is T {
  if (!pred(x)) {
    throw new AssertError(
      options.message ?? assertMessageFactory(x, pred, options.name),
    );
  }
}
