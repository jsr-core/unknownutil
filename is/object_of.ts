import type { FlatType } from "../_typeutil.ts";
import { rewriteName } from "../_funcutil.ts";
import {
  annotate,
  type WithOptional,
  type WithPredObj,
  type WithReadonly,
} from "../_annotation.ts";
import type { Predicate } from "../type.ts";

/**
 * Return a type predicate function that returns `true` if the type of `x` is `ObjectOf<T>`.
 *
 * To enhance performance, users are advised to cache the return value of this function and mitigate the creation cost.
 *
 * If {@linkcode asOptional} is specified in the predicate function, the property becomes optional.
 * If {@linkcode asReadonly} is specified in the predicate function, the property becomes readonly.
 *
 * The number of keys of `x` must be greater than or equal to the number of keys of `predObj`.
 * Use {@linkcode isStrictOf} if you want to check the exact number of keys.
 *
 * ```ts
 * import { as, is } from "@core/unknownutil";
 *
 * const isMyType = is.ObjectOf({
 *   a: is.Number,
 *   b: is.String,
 *   c: as.Optional(is.Boolean),
 *   d: as.Readonly(is.String),
 * });
 * const a: unknown = { a: 0, b: "a", d: "d" };
 * if (isMyType(a)) {
 *   const _: { a: number; b: string; c?: boolean | undefined, readonly d: string } = a;
 * }
 * ```
 */
export function isObjectOf<
  T extends Record<PropertyKey, Predicate<unknown>>,
>(predObj: T): Predicate<ObjectOf<T>> & WithPredObj<T> {
  return annotate(
    rewriteName(
      (x: unknown): x is ObjectOf<T> => {
        if (
          x == null ||
          typeof x !== "object" && typeof x !== "function" ||
          Array.isArray(x)
        ) return false;
        // Check each values
        for (const k in predObj) {
          if (!predObj[k]((x as T)[k])) return false;
        }
        return true;
      },
      "isObjectOf",
      predObj,
    ),
    "predObj",
    predObj,
  );
}

type ObjectOf<T extends Record<PropertyKey, Predicate<unknown>>> = FlatType<
  // Readonly/Optional
  & {
    readonly [
      K in keyof T as T[K] extends WithReadonly
        ? T[K] extends WithOptional ? K : never
        : never
    ]?: T[K] extends Predicate<infer U> ? U : never;
  }
  // Readonly/Non optional
  & {
    readonly [
      K in keyof T as T[K] extends WithReadonly
        ? T[K] extends WithOptional ? never : K
        : never
    ]: T[K] extends Predicate<infer U> ? U : never;
  }
  // Non readonly/Optional
  & {
    [
      K in keyof T as T[K] extends WithReadonly ? never
        : T[K] extends WithOptional ? K
        : never
    ]?: T[K] extends Predicate<infer U> ? U : never;
  }
  // Non readonly/Non optional
  & {
    [
      K in keyof T as T[K] extends WithReadonly ? never
        : T[K] extends WithOptional ? never
        : K
    ]: T[K] extends Predicate<infer U> ? U : never;
  }
>;
