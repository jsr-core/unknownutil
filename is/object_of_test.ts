import { assertEquals } from "@std/assert";
import { assertSnapshot } from "@std/testing/snapshot";
import { assertType } from "@std/testing/types";
import type { Equal } from "../_testutil.ts";
import { is } from "./mod.ts";
import { as } from "../as/mod.ts";
import { isObjectOf } from "./object_of.ts";

Deno.test("isObjectOf<T>", async (t) => {
  const predObj = {
    a: is.Number,
    b: is.String,
    c: is.Boolean,
  };

  await t.step("returns properly named predicate function", async (t) => {
    assertEquals(typeof isObjectOf({}), "function");
    await assertSnapshot(t, isObjectOf(predObj).name);
    await assertSnapshot(
      t,
      isObjectOf({ a: (_x): _x is string => false }).name,
    );
    await assertSnapshot(
      t,
      isObjectOf({ a: isObjectOf({ b: isObjectOf({ c: is.Boolean }) }) }).name,
    );
  });

  await t.step("returns true on T object", () => {
    assertEquals(isObjectOf(predObj)({ a: 0, b: "a", c: true }), true);
    assertEquals(
      isObjectOf(predObj)({ a: 0, b: "a", c: true, d: "ignored" }),
      true,
      "Undefined properties are ignored",
    );
    assertEquals(
      isObjectOf(predObj)(
        Object.assign(() => void 0, { a: 0, b: "a", c: true }),
      ),
      true,
      "Function are treated as an object",
    );
  });

  await t.step("returns false on non T object", () => {
    assertEquals(isObjectOf(predObj)("a"), false, "Value is not an object");
    assertEquals(
      isObjectOf(predObj)({ a: 0, b: "a", c: "" }),
      false,
      "Object have a different type property",
    );
    assertEquals(
      isObjectOf(predObj)({ a: 0, b: "a" }),
      false,
      "Object does not have one property",
    );
    assertEquals(
      isObjectOf({ 0: is.String })(["a"]),
      false,
      "Value is not an object",
    );
  });

  await t.step("returns true on T instance", () => {
    const date = new Date();
    const predObj = {
      getFullYear: is.Function,
    };
    assertEquals(isObjectOf(predObj)(date), true);
  });

  await t.step("predicated type is correct", () => {
    const predObj2 = {
      a: as.Readonly(as.Optional(is.String)),
      b: as.Optional(as.Readonly(is.String)),
      c: as.Readonly(is.String),
      d: as.Optional(is.String),
      e: as.Unreadonly(as.Unoptional(as.Readonly(as.Optional(is.String)))),
      f: as.Unoptional(as.Unreadonly(as.Optional(as.Readonly(is.String)))),
    };
    const a: unknown = { a: 0, b: "a", c: true };

    if (isObjectOf(predObj)(a)) {
      assertType<Equal<typeof a, { a: number; b: string; c: boolean }>>(true);
    }

    if (isObjectOf(predObj2)(a)) {
      assertType<
        Equal<
          typeof a,
          {
            readonly a?: string;
            readonly b?: string;
            readonly c: string;
            d?: string;
            e: string;
            f: string;
          }
        >
      >(true);
    }
  });

  await t.step(
    "does not affect prototype properties of 'predObj'",
    async (t) => {
      const prototypeObj = {
        a: is.Number,
        b: is.Boolean,
      };
      // deno-lint-ignore ban-types
      const predObj = Object.assign(Object.create(prototypeObj) as {}, {
        c: is.String,
      });

      await t.step("returns true on T object", () => {
        assertEquals(isObjectOf(predObj)({ c: "a" }), true);
        assertEquals(
          isObjectOf(predObj)({ a: "ignored", b: "ignored", c: "a" }),
          true,
          "Predicates defined in the prototype are ignored",
        );
        assertEquals(
          isObjectOf(predObj)({ c: "a", d: "ignored" }),
          true,
          "Undefined properties are ignored",
        );
        assertEquals(
          isObjectOf(predObj)(Object.assign(() => void 0, { c: "a" })),
          true,
          "Function are treated as an object",
        );
      });

      await t.step("returns false on non T object", () => {
        assertEquals(isObjectOf(predObj)("a"), false, "Value is not an object");
        assertEquals(
          isObjectOf(predObj)({ a: 0, b: true, c: 1 }),
          false,
          "Object have a different type property",
        );
        assertEquals(
          isObjectOf(predObj)({ a: 0, b: true }),
          false,
          "Object does not have one property",
        );
        assertEquals(
          isObjectOf({ 0: is.String })(["a"]),
          false,
          "Value is not an object",
        );
      });
    },
  );

  await t.step("with symbol properties", async (t) => {
    const s = Symbol("s");
    const predObj = {
      a: is.Number,
      b: is.String,
      [s]: is.Boolean,
    };

    await t.step("returns properly named predicate function", async (t) => {
      await assertSnapshot(t, isObjectOf(predObj).name);
      await assertSnapshot(
        t,
        isObjectOf({
          [Symbol("a")]: isObjectOf({
            [Symbol("b")]: isObjectOf({ [Symbol("c")]: is.Boolean }),
          }),
        }).name,
      );
    });

    await t.step("returns true on T object", () => {
      assertEquals(isObjectOf(predObj)({ a: 0, b: "a", [s]: true }), true);
      assertEquals(
        isObjectOf(predObj)({ a: 0, b: "a", [s]: true, d: "ignored" }),
        true,
        "Undefined properties are ignored",
      );
      assertEquals(
        isObjectOf(predObj)({
          a: 0,
          b: "a",
          [s]: true,
          [Symbol("t")]: "ignored",
        }),
        true,
        "Undefined symbol properties are ignored",
      );
      assertEquals(
        isObjectOf(predObj)(
          Object.assign(() => void 0, { a: 0, b: "a", [s]: true }),
        ),
        true,
        "Function are treated as an object",
      );
    });

    await t.step("returns false on non T object", () => {
      assertEquals(isObjectOf(predObj)("a"), false, "Value is not an object");
      assertEquals(
        isObjectOf(predObj)({ a: 0, b: "a", [s]: "" }),
        false,
        "Object have a different type symbol property",
      );
      assertEquals(
        isObjectOf(predObj)({ a: 0, b: "a" }),
        false,
        "Object does not have symbol property",
      );
      const arrayWithSymbolProp = ["ignored"];
      // deno-lint-ignore no-explicit-any
      (arrayWithSymbolProp as any)[s] = true;
      assertEquals(
        isObjectOf({ [s]: is.Boolean })(arrayWithSymbolProp),
        false,
        "Value is not an object",
      );
    });

    await t.step("predicated type is correct", () => {
      const a = Symbol("a");
      const b = Symbol("b");
      const c = Symbol("c");
      const d = Symbol("d");
      const e = Symbol("e");
      const f = Symbol("f");
      const predObj2 = {
        [a]: as.Readonly(as.Optional(is.String)),
        [b]: as.Optional(as.Readonly(is.String)),
        [c]: as.Readonly(is.String),
        [d]: as.Optional(is.String),
        [e]: as.Unreadonly(as.Unoptional(as.Readonly(as.Optional(is.String)))),
        [f]: as.Unoptional(as.Unreadonly(as.Optional(as.Readonly(is.String)))),
      };
      const x: unknown = {};

      if (isObjectOf(predObj)(x)) {
        assertType<
          Equal<
            typeof x,
            {
              a: number;
              b: string;
              [s]: boolean;
            }
          >
        >(true);
      }

      if (isObjectOf(predObj2)(x)) {
        assertType<
          Equal<
            typeof x,
            {
              readonly [a]?: string;
              readonly [b]?: string;
              readonly [c]: string;
              [d]?: string;
              [e]: string;
              [f]: string;
            }
          >
        >(true);
      }
    });
  });
});
