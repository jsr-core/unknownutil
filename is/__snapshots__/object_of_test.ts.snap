export const snapshot = {};

snapshot[`isObjectOf<T> > returns properly named predicate function 1`] = `
"isObjectOf({
  a: isNumber,
  b: isString,
  c: isBoolean
})"
`;

snapshot[`isObjectOf<T> > returns properly named predicate function 2`] = `"isObjectOf({a: a})"`;

snapshot[`isObjectOf<T> > returns properly named predicate function 3`] = `
"isObjectOf({
  a: isObjectOf({
    b: isObjectOf({c: isBoolean})
  })
})"
`;

snapshot[`isObjectOf<T> > with symbol properties > returns properly named predicate function 1`] = `
"isObjectOf({
  a: isNumber,
  b: isString,
  Symbol(s): isBoolean
})"
`;

snapshot[`isObjectOf<T> > with symbol properties > returns properly named predicate function 2`] = `
"isObjectOf({
  Symbol(a): isObjectOf({
    Symbol(b): isObjectOf({Symbol(c): isBoolean})
  })
})"
`;
